backend/src/main.ts

```diff
@@ -1,27 +1,49 @@
-import * as dotenv from "dotenv";
-dotenv.config();
-
 import { createServer } from "http";
-import { Server } from "socket.io";
+import { Server, Socket } from "socket.io";
+import { shareReplay, Subject, take, timeout, } from "rxjs";
+
+import { CoffeeCall, Command, TimedCommand } from "./models";
+import { getAllowedCorsOrigins, getServerListeningPort } from "./environment";
+import { convertCommandsToCalls } from "./message-pipeline";
+
+type VoidListener<T> = (event: T) => void;
+type CoffeeCallerEventsMap = { coffee: VoidListener<TimedCommand>, coffeeCalls: VoidListener<CoffeeCall> };
 
 const httpServer = createServer();
-const io = new Server(httpServer, {
+const io = new Server<CoffeeCallerEventsMap>(httpServer, {
     cors: {
-        origin: process.env.CORS_ALLOWED_ORIGIN || 'http://localhost:3000',
+        origin: getAllowedCorsOrigins(),
         methods: ["GET", "POST"]
     }
 });
 
+const messageSubject = new Subject<Command>();
+const processedCoffeeCalls = messageSubject.pipe(convertCommandsToCalls, shareReplay(1));
+
+function broadcastCommand(command: Command) {
+    const timed = { ...command, broadcastAt: new Date() };
+    console.log('→ broadcast command msg:', timed);
+    io.sockets.emit("coffee", timed);
+}
+
+function logCommand(command: Command) {
+    console.log('← received command msg:', command);
+}
+
+function putCommandIntoPipeline(command: Command) {
+    messageSubject.next(command);
+}
+
 io.on("connection", (socket) => {
     console.log(`Got new connection: ${socket.id}`);
-    socket.on("coffee", (data) => {
-        io.sockets.emit("coffee", JSON.stringify({
-            ...JSON.parse(data),
-            broadcastAt: new Date().toISOString()
-        }));
-    });
+
+    socket.on("coffee", logCommand);
+    socket.on("coffee", broadcastCommand)
+    socket.on("coffee", putCommandIntoPipeline);
+
+    processedCoffeeCalls.subscribe((coffeeCall) => socket.emit("coffeeCalls", coffeeCall));
```

Comment:
Suggestion to rename the message channels (events?):
- `coffee` → `coffeeRequest` (coming from clients to the BE: `join`, `leave`, `start`)
- `coffeeCall` keep as it is (sent from BE to clients whenever something changes)